[SPARK-32858][SQL] UnwrapCastInBinaryComparison: support other numeric types

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

In SPARK-24994 we implemented unwrapping cast for **integral types**. This extends it to support **numeric types** such as float/double/decimal, so that filters involving these types can be better pushed down to data sources.

Unlike the cases of integral types, conversions between numeric types can result to rounding up or downs. Consider the following case:

```sql
cast(e as double) < 1.9
```

assume type of `e` is short, since 1.9 is not representable in the type, the casting will either truncate or round. Now suppose the literal is truncated, we cannot convert the expression to:

```sql
e < cast(1.9 as short)
```

as in the previous implementation, since if `e` is 1, the original expression evaluates to true, but converted expression will evaluate to false.

To resolve the above, this PR first finds out whether casting from the wider type to the narrower type will result to truncate or round, by comparing a _roundtrip value_ derived from **converting the literal first to the narrower type, and then to the wider type**, versus the original literal value. For instance, in the above, we'll first obtain a roundtrip value via the conversion (double) 1.9 -> (short) 1 -> (double) 1.0, and then compare it against 1.9. 

<img width="1153" alt="Screen Shot 2020-09-28 at 3 30 27 PM" src="https://user-images.githubusercontent.com/506679/94492719-bd29e780-019f-11eb-9111-71d6e3d157f7.png">
                                   
Now in the case of truncate, we'd convert the original expression to:
```sql
e <= cast(1.9 as short)
```
instead, so that the conversion also is valid when `e` is 1.

For more details, please check [this blog post](https://prestosql.io/blog/2019/05/21/optimizing-the-casts-away.html) by Presto which offers a very good explanation on how it works.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

For queries such as:
```sql
SELECT * FROM tbl WHERE short_col < 100.5
```
The predicate `short_col < 100.5` can't be pushed down to data sources because it involves casts. This eliminates the cast so these queries can run more efficiently.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Unit tests

sunchao committed:
[SPARK-32858][SQL] UnwrapCastInBinaryComparison: support other numeric types

sunchao committed:
More test coverage

SparkQA commented:
**[Test build #128838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128838/testReport)** for PR 29792 at commit [`65833e7`](https://github.com/apache/spark/commit/65833e7d38a98495ec898bae2ec33adef7054aa5).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33461/
Test PASSed.

SparkQA commented:
**[Test build #128838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128838/testReport)** for PR 29792 at commit [`65833e7`](https://github.com/apache/spark/commit/65833e7d38a98495ec898bae2ec33adef7054aa5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128838/
Test PASSed.

dbtsai commented:
cc @viirya @dongjoon-hyun @cloud-fan @gatorsmile

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:
```diff
@@ -67,6 +77,23 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
     assertEquivalent(castInt(f) < v.toInt, falseIfNotNull(f))
+
+    val d = Float.NegativeInfinity
+    assertEquivalent(castDouble(f2) > d.toDouble, f2 =!= d)
```
is casting double to float rounding up or rounding down?

sunchao replied:
it is rounding down, see below for a test on this.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:
```diff
@@ -79,13 +106,65 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
+    assertEquivalent(castDouble(f) <= doubleValue, f <= doubleValue.toShort)
+    assertEquivalent(castDouble(f) < doubleValue, f <= doubleValue.toShort)
+
+    // Cases for rounding up: 3.14 will be rounded to 3.14000010... after casting to float
```
so casting double to float can be either rounding up or down, depend on the value?

sunchao replied:
@cloud-fan Sorry i was wrong in the above comment (somehow I was thinking casting from double to short there). 

Yes, it appears that casting from double to float can be either rounding up or down, depending on value:

```scala
scala> val x = 0.39999999
x: Double = 0.39999999

scala> val y = x.toFloat
y: Float = 0.39999998

scala > val x = 0.49999999
y: Double = 0.49999999

scala> val y = x.toFloat
y: Float = 0.5
```
Also the test here is incorrect. casting 3.14 of double to float doesn't change anything. I'll update it.

cloud-fan replied:
This is an important point. Can we explain how to know it's rounding up or down in the PR description?

sunchao replied:
Yup, will do. This is a good point.

sunchao replied:
@cloud-fan updated the description. Please take another look. Thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -35,18 +35,34 @@ import org.apache.spark.sql.types._
+ *  - `cast(fromExp, toType) >= value` ==> `fromExp >= cast(value, fromType)`
+ *  - `cast(fromExp, toType) === value` ==> if(isnull(fromExp), null, false)
+ *  - `cast(fromExp, toType) <=> value` ==> false (if `fromExp` is deterministic)
+ *  - `cast(fromExp, toType) <=> value` ==> `cast(fromExp, toType) <=> value` (if `fromExp` is
```
We can remove this because the rule does nothing for it.

sunchao replied:
Thanks, will do.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -200,25 +252,27 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+    case ShortType => Some((Short.MinValue, Short.MaxValue))
+    case IntegerType => Some((Int.MinValue, Int.MaxValue))
+    case LongType => Some((Long.MinValue, Long.MaxValue))
+    case FloatType => Some((Float.NegativeInfinity, Float.NaN))
```
why the upper bound is not `PositiveInfinity`?

sunchao replied:
This is because `PositiveInfinity` is considered to be < `NaN` in Spark. If we treat it as the upper bound, rules handling the upper bounds will not be valid. For instance the following expr:
```sql
cast(e as double) > double('+inf')
```
would be converted to
```sql
e === double('+inf')
```
which won't be correct if `e` evaluates to `double('NaN')`.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -116,82 +132,118 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+    }
+
+    // When we reach to this point, it means either there is no min/max for the `fromType` (e.g.,
+    // decimal type), or that the literal `value` is within range `(min, max)`. For these, we
```
why it's safe to skip range check for decimal type?

sunchao replied:
It is safe since knowing min/max for a type just gives us more opportunity for optimizations. I skipped decimal type here because (it seems) there is no min/max defined in the `DecimalType`, unlike other numeric types.

cloud-fan replied:
makes sense. Can you give an example of how the range of float/double helps the optimization?

cloud-fan replied:
is it only useful to `NaN`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -116,82 +132,118 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+      // narrower type. In this case we simply return the original expression.
+      return exp
+    }
+    val valueRoundTrip = Cast(Literal(newValue, fromType), toType).eval()
```
The case I'm worried about is `cast(float_col as double) cmp double_lit`. It's not straightforward to me that a `double -> float -> double` roundtrip can tell rounding up or down. is it because `float -> double` can only be rounding up?

sunchao replied:
So double to float can result to either rounding up or down. For instance, by casting 3.14 in double to float, even though the value is still 3.14, the binary representation is rounded up:

3.14 in double:
```
0 10000000000 1001 0001 1110 1011 1000 0101 0001 1110 1011 1000 0101 0001 1111
```

3.14 in float
```
0 10000000 1001 0001 1110 1011 1000 011
```
Here the sign bit and exponent bits (11 and 8 bits respectively for double and float) are the same for both float and double. However, in the fraction part, the last is rounded up to 1.

After casting back to double, there won't be any rounding up or down - the remaining digits are simply padded with 0:
```
0 10000000000 1001 0001 1110 1011 1000 0110 0000000000000000000000000000
```

dbtsai replied:
Is it defined as part of IEEE Standard for Floating-Point Arithmetic (IEEE 754)?

sunchao replied:
Yes I think both the binary format as well as rounding rules are specified in IEEE 754. There are a few rounding rules and I think the default one is "rounding to half even".

sunchao committed:
Minor refactoring

SparkQA commented:
**[Test build #129311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129311/testReport)** for PR 29792 at commit [`3340de4`](https://github.com/apache/spark/commit/3340de447c551ca2c7d0e9a470b99a02a87913b1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33926/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33926/
Test PASSed.

SparkQA commented:
**[Test build #129311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129311/testReport)** for PR 29792 at commit [`3340de4`](https://github.com/apache/spark/commit/3340de447c551ca2c7d0e9a470b99a02a87913b1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129311/
Test PASSed.

sunchao commented:
ping @cloud-fan - addressed your comments, could you take another look at this? thanks!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -116,82 +128,118 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+    val newValue = Cast(Literal(value), fromType).eval()
+    if (newValue == null) {
+      // This means the cast failed, for instance, due to the value is not representable in the
+      // narrower type. In this case we simply return the original expression.
```
can you give a real example here?

cloud-fan replied:
I see, it's for decimal only. It's better to make the comment more explicit.

sunchao replied:
yup will do - there is also a test case covering this.

cloud-fan commented:
The patch LGTM. Can we have an end-to-end test suite for it? The current tests prove that the optimized expression tree is what we expect, it's better to have high-level tests to prove that, after optimization, the query still returns corrected result.

sunchao commented:
> Can we have an end-to-end test suite for it?

Thanks - this is a good suggestion. Will add that.

sunchao committed:
Add end-to-end test suite

SparkQA commented:
**[Test build #129566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129566/testReport)** for PR 29792 at commit [`94942bf`](https://github.com/apache/spark/commit/94942bf2042a5c179ce9eda7fb90493636905078).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34172/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34172/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34172/
Test PASSed.

SparkQA commented:
**[Test build #129566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129566/testReport)** for PR 29792 at commit [`94942bf`](https://github.com/apache/spark/commit/94942bf2042a5c179ce9eda7fb90493636905078).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class UnwrapCastInComparisonEndToEndSuite extends QueryTest with SharedSparkSession `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129566/
Test PASSed.

sunchao commented:
@cloud-fan added an e2e test suite. Please take another look, thanks.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.Decimal
+
+class UnwrapCastInComparisonEndToEndSuite extends QueryTest with SharedSparkSession {
```
Could you add these end-2-end tests in `SQLQueryTestSuite` instead of making a new suite?

sunchao replied:
Yea I can add the test there instead - I was just following the existing [ReplaceNullWithFalseInPredicateEndToEndSuite](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/ReplaceNullWithFalseInPredicateEndToEndSuite.scala) though.

cloud-fan replied:
I think it's fine to follow ReplaceNullWithFalseInPredicateEndToEndSuite here.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -35,18 +35,32 @@ import org.apache.spark.sql.types._
+ *  - `cast(fromExp, toType) <= value` ==> `fromExp < cast(value, fromType)`
+ *  - `cast(fromExp, toType) < value` ==> `fromExp < cast(value, fromType)`
+ *
+ *  Similarly for the case when casting `value` to `fromType` causes rounding down.
```
nit: wrong indent.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:
```diff
@@ -200,25 +248,27 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+    case IntegerType => Some((Int.MinValue, Int.MaxValue))
+    case LongType => Some((Long.MinValue, Long.MaxValue))
+    case FloatType => Some((Float.NegativeInfinity, Float.NaN))
+    case DoubleType => Some((Double.NegativeInfinity, Double.NaN))
```
Looks it does not have any test for this code path, so could you add some tests for it. (NOTE: I think `byte`, `int`, and `long` are not tested in `UnwrapCastInBinaryComparisonSuite`, too)

sunchao replied:
Will add a test case (although I think it will be pretty trivial). I only added tests for `short` in the previous PR because the handling for other integral types is exactly the same.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+    val t = "test_table"
+    withTable(t) {
+      Seq[(Integer, java.lang.Short, java.lang.Float)](
+        (1, 100.toShort, 3.14.toFloat), (2, Short.MaxValue, Float.NaN), (3, null, null))
```
can we test `Float.PositiveInfinity` as well?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+      checkAnswer(df.select("c1").where(s"c3 <= double('nan')"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c3 < double('nan')"), Row(1))
+
+      lit = positiveInt
```
this doesn't match the test case name `cases when literal is max`

We can put it in a new test `cases when literal exceeds max`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+      checkAnswer(df.select("c1").where(s"c3 <= double('-inf')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 < double('-inf')"), Seq.empty)
+
+      lit = negativeInt
```
ditto

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+    }
+  }
+
+  test("cases when literal is within range (min, max) and rounding for ") {
```
`rounding for `: uncompleted sentence?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala:
```diff
@@ -0,0 +1,165 @@
+      val df = spark.table(t)
+
+      var lit = Short.MinValue.toInt
+      checkAnswer(df.select("c1").where(s"c2 > $lit"), Row(1))
```
can we put the `where` before `select`? It's weird to see we select only `c1` and then filter on `c2`.

sunchao committed:
Address comments

- Added test case for getRange()
- Added test for Float.PositiveInfinity and Float.MinValue/Float/MaxValue
- Move `select` after `where`
- Separate test cases
- Fix indentation

SparkQA commented:
**[Test build #129704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129704/testReport)** for PR 29792 at commit [`76b9f73`](https://github.com/apache/spark/commit/76b9f73862380dda40b332b56464278bfb0b88d3).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34310/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34310/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34310/
Test PASSed.

SparkQA commented:
**[Test build #129704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129704/testReport)** for PR 29792 at commit [`76b9f73`](https://github.com/apache/spark/commit/76b9f73862380dda40b332b56464278bfb0b88d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129704/
Test PASSed.

sunchao commented:
Addressed comments. Please take another look @cloud-fan @maropu . Thanks!

cloud-fan commented:
thanks, merging to master!
